#!/usr/bin/env nextflow
/*
========================================================================================
                         nf-core/hgtsimulation
========================================================================================
 nf-core/hgtsimulation Analysis Pipeline.
 #### Homepage / Documentation
 https://github.com/nf-core/hgtsimulation
----------------------------------------------------------------------------------------
*/

def helpMessage() {
    // TODO nf-core: Add to this help message with new command line parameters
    log.info nfcoreHeader()
    log.info"""

    Usage:

    The typical command for running the pipeline is as follows:

    nextflow run nf-core/hgtsimulation --reads '*_R{1,2}.fastq.gz' -profile docker

    Mandatory arguments:
      --recipients                  Path to input data (must be surrounded with quotes) of the recipient genomes.
      --plasmids                    Path to plasmid FASTA files.
      -profile                      Configuration profile to use. Can use multiple (comma separated)
                                    Available: conda, docker, singularity, awsbatch, test and more.

    Options:
      --num_reads                   Number of reads to simulate per sample.
      --read_length                 Length of simulated reads
      --min_insert                  Minimum simulated insert length
      --max_insert                  Maximum simulated insert length
      --adapter1                    Simulated reads adapter sequence for read 1
      --adapter2                    Simulated reads adapter sequence for read 2
      --unicycler_args              Arguments for Unicycler. Must be a string.

    Other options:
      --outdir                      The output directory where the results will be saved
      --email                       Set this parameter to your e-mail address to get a summary e-mail with details of the run sent to you when the workflow exits
      --email_on_fail               Same as --email, except only send mail if the workflow is not successful
      --maxMultiqcEmailFileSize     Theshold size for MultiQC report to be attached in notification email. If file generated by pipeline exceeds the threshold, it will not be attached (Default: 25MB)
      -name                         Name for the pipeline run. If not specified, Nextflow will automatically generate a random mnemonic.

    AWSBatch options:
      --awsqueue                    The AWSBatch JobQueue that needs to be set when running on AWSBatch
      --awsregion                   The AWS Region for your AWS Batch job to run on
    """.stripIndent()
}

// Show help message
if (params.help) {
    helpMessage()
    exit 0
}

/*
 * SET UP CONFIGURATION VARIABLES
 */


// Has the run name been specified by the user?
//  this has the bonus effect of catching both -name and --name
custom_runName = params.name
if (!(workflow.runName ==~ /[a-z]+_[a-z]+/)) {
  custom_runName = workflow.runName
}

if ( workflow.profile == 'awsbatch') {
  // AWSBatch sanity checking
  if (!params.awsqueue || !params.awsregion) exit 1, "Specify correct --awsqueue and --awsregion parameters on AWSBatch!"
  // Check outdir paths to be S3 buckets if running on AWSBatch
  // related: https://github.com/nextflow-io/nextflow/issues/813
  if (!params.outdir.startsWith('s3:')) exit 1, "Outdir not on S3 - specify S3 Bucket to run on AWSBatch!"
  // Prevent trace files to be stored on S3 since S3 does not support rolling files.
  if (workflow.tracedir.startsWith('s3:')) exit 1, "Specify a local tracedir or run without trace! S3 cannot be used for tracefiles."
}

// Stage config files
ch_multiqc_config = file(params.multiqc_config, checkIfExists: true)
ch_output_docs = file("$baseDir/docs/output.md", checkIfExists: true)

/*
 * Create a channel for input FASTA files
 */
recipients_ch = Channel.fromPath(params.recipients)
plasmids_ch = Channel.fromPath(params.plasmids)

// Header log info
log.info nfcoreHeader()
def summary = [:]
if (workflow.revision) summary['Pipeline Release'] = workflow.revision
summary['Run Name']         = custom_runName ?: workflow.runName
// TODO nf-core: Report custom parameters here
summary['Fasta']            = params.fasta
summary['Plasmids']        = params.plasmids
summary['Max Resources']    = "$params.max_memory memory, $params.max_cpus cpus, $params.max_time time per job"
if (workflow.containerEngine) summary['Container'] = "$workflow.containerEngine - $workflow.container"
summary['Output dir']       = params.outdir
summary['Launch dir']       = workflow.launchDir
summary['Working dir']      = workflow.workDir
summary['Script dir']       = workflow.projectDir
summary['User']             = workflow.userName
if (workflow.profile == 'awsbatch') {
  summary['AWS Region']     = params.awsregion
  summary['AWS Queue']      = params.awsqueue
}
summary['Config Profile'] = workflow.profile
if (params.config_profile_description) summary['Config Description'] = params.config_profile_description
if (params.config_profile_contact)     summary['Config Contact']     = params.config_profile_contact
if (params.config_profile_url)         summary['Config URL']         = params.config_profile_url
if (params.email || params.email_on_fail) {
  summary['E-mail Address']    = params.email
  summary['E-mail on failure'] = params.email_on_fail
  summary['MultiQC maxsize']   = params.maxMultiqcEmailFileSize
}
log.info summary.collect { k,v -> "${k.padRight(18)}: $v" }.join("\n")
log.info "-\033[2m--------------------------------------------------\033[0m-"

// Check the hostnames against configured profiles
checkHostname()

def create_workflow_summary(summary) {
    def yaml_file = workDir.resolve('workflow_summary_mqc.yaml')
    yaml_file.text  = """
    id: 'nf-core-hgtsimulation-summary'
    description: " - this information is collected when the pipeline is started."
    section_name: 'nf-core/hgtsimulation Workflow Summary'
    section_href: 'https://github.com/nf-core/hgtsimulation'
    plot_type: 'html'
    data: |
        <dl class=\"dl-horizontal\">
${summary.collect { k,v -> "            <dt>$k</dt><dd><samp>${v ?: '<span style=\"color:#999999;\">N/A</a>'}</samp></dd>" }.join("\n")}
        </dl>
    """.stripIndent()

   return yaml_file
}

/*
 * Parse software version numbers
 */
process get_software_versions {
    publishDir "${params.outdir}/pipeline_info", mode: 'copy',
        saveAs: { filename ->
            if (filename.indexOf(".csv") > 0) filename
            else null
        }

    output:
    file 'software_versions_mqc.yaml' into software_versions_yaml
    file "software_versions.csv"

    script:
    // TODO nf-core: Get all tools to print their version number here
    """
    echo $workflow.manifest.version > v_pipeline.txt
    echo $workflow.nextflow.version > v_nextflow.txt
    fastqc --version > v_fastqc.txt
    multiqc --version > v_multiqc.txt
    scrape_software_versions.py &> software_versions_mqc.yaml
    """
}

/*
 * STEP 1 - CONCATENATE FASTA AND PLASMID
 */
process concatenate_fasta_plasmid {
    publishDir "${params.outdir}/${plasmid.getSimpleName()}/${recipient.getSimpleName()}", mode: 'copy'

    input:
    each path(recipient) from recipients_ch
    each path(plasmid) from plasmids_ch

    output:
    tuple(val ("${plasmid.getSimpleName()}"), val ("${recipient.getSimpleName()}"), path("${plasmid.getSimpleName()}_${recipient.getSimpleName()}.fasta")) into readsim_ch

    script:
    """
    cat ${recipient} ${plasmid} > ${plasmid.getSimpleName()}_${recipient.getSimpleName()}.fasta
    """
}

/* 
 * STEP 2 - SIMULATE READS
 */
process simulate_reads {
    publishDir "${params.outdir}/${plasmid}/${recipient}/simulated_reads", mode: 'copy'

    input:
    tuple(plasmid, recipient, path(fasta)) from readsim_ch

    output:
    tuple(plasmid, recipient, path("${fasta.getSimpleName()}_*.fastq.gz")) into fastp_ch
    
    script:
    """
    randomreads.sh ref=${fasta} out1=${fasta.getSimpleName()}_1.fastq.gz out2=${fasta.getSimpleName()}_2.fastq.gz reads=${params.num_reads} length=${params.read_length} illuminanames=t paired=t mininsert=${params.min_insert} maxinsert=${params.max_insert} fragadapter=${params.adapter1} fragadapter2=${params.adapter2}
    """
}

/*
 * STEP 3 - FILTER
 */
process fastp {
    publishDir "${params.outdir}/${plasmid}/${recipient}/fastp", mode: 'copy'

    input:
    tuple(plasmid, recipient, path(reads)) from fastp_ch

    output:
    tuple(plasmid, recipient, path("trim*.fastq.gz")) into assembly_ch
    path("*fastp.{json,html}")

    script:
    """
    fastp -i ${reads[0]} -I ${reads[1]} -o trim_${reads[0]} -O trim_${reads[1]} -w ${task.cpus} --json ${plasmid}_${recipient}_fastp.json --html ${plasmid}_${recipient}_fastp.html
    """
}

/*
 * STEP 3 - ASSEMBLE
 */
process unicycler {
    tag "${plasmid}_${recipient}"
    publishDir "${params.outdir}/${plasmid}/${recipient}", mode: 'copy'

    input:
    tuple(plasmid, recipient, path(reads)) from assembly_ch

    output:
    tuple (plasmid, recipient, path("${plasmid}_${recipient}_assembly.fasta"), path("${plasmid}_${recipient}_assembly.gfa")) into (quast_ch, abricate_ch)
    path("${plasmid}_${recipient}_assembly.gfa")
    path("${plasmid}_${recipient}_assembly.png")
    path("${plasmid}_${recipient}_unicycler.log")

    script:
    """
    unicycler --threads ${task.cpus} ${params.unicycler_args} --keep 0 -o . -1 ${reads[0]} -2 ${reads[1]}
    mv unicycler.log ${plasmid}_${recipient}_unicycler.log
    mv assembly.gfa ${plasmid}_${recipient}_assembly.gfa
    mv assembly.fasta ${plasmid}_${recipient}_assembly.fasta
    Bandage image ${plasmid}_${recipient}_assembly.gfa ${plasmid}_${recipient}_assembly.png
    """
}

/*
 * STEP 4 - QUAST
 */
process quast {
    tag "${plasmid}_${recipient}"
    publishDir "${params.outdir}/${plasmid}/${recipient}/quast", mode: 'copy'

    input:
    tuple (plasmid, recipient, path(assembly), path(gfa)) from quast_ch
    
    output:
    path("${plasmid}_${recipient}_assembly_QC/")
    path("${plasmid}_${recipient}_assembly_QC/report.tsv") into quast_log_ch
    path("v_quast.txt") into quast_version_ch

    script:
    """
    quast -t ${task.cpus} -o ${plasmid}_${recipient}_assembly_QC ${assembly}
    quast -v > v_quast.txt
    """
}

/*
 * STEP 5 - ABRICATE
 */
process abricate {
    tag "${plasmid}_${recipient}"
    publishDir "${params.outdir}/${plasmid}/${recipient}/abricate", mode: 'copy'

    input:
    tuple(plasmid, recipient, path(assembly), path(gfa)) from abricate_ch
    
    output:
    tuple(plasmid, recipient, path(assembly), path(gfa), path("${plasmid}_${recipient}_assembly.genes")) into amr_ch

    script:
    """
    abricate --threads ${task.cpus} --db ncbi ${assembly} > ${plasmid}_${recipient}_assembly.amr
    cat ${plasmid}_${recipient}_assembly.amr | awk 'FNR > 1 {print \$6}' > ${plasmid}_${recipient}_assembly.genes
    """
    
}

/* 
 * STEP 6 - ORGANIZE BY GENE
 */
process sort_genes {
    publishDir "${params.outdir}", mode: 'copy'

    input:
    tuple(plasmid, recipient, path(assembly), path(gfa), path(genes)) from amr_ch 
    
    output:
    path("*/")
    path("*/*.{fasta,gfa}")

    script:
    """
    while read p
    do
    mkdir $p
    cp $assembly $p
    cp $gfa $p
    done < $genes
    """
}

/*
 * STEP 3 - Output Description HTML
 */
process output_documentation {
    publishDir "${params.outdir}/pipeline_info", mode: 'copy'

    input:
    file output_docs from ch_output_docs

    output:
    file "results_description.html"

    script:
    """
    markdown_to_html.r $output_docs results_description.html
    """
}

/*
 * Completion e-mail notification
 */
workflow.onComplete {

    // Set up the e-mail variables
    def subject = "[nf-core/hgtsimulation] Successful: $workflow.runName"
    if (!workflow.success) {
      subject = "[nf-core/hgtsimulation] FAILED: $workflow.runName"
    }
    def email_fields = [:]
    email_fields['version'] = workflow.manifest.version
    email_fields['runName'] = custom_runName ?: workflow.runName
    email_fields['success'] = workflow.success
    email_fields['dateComplete'] = workflow.complete
    email_fields['duration'] = workflow.duration
    email_fields['exitStatus'] = workflow.exitStatus
    email_fields['errorMessage'] = (workflow.errorMessage ?: 'None')
    email_fields['errorReport'] = (workflow.errorReport ?: 'None')
    email_fields['commandLine'] = workflow.commandLine
    email_fields['projectDir'] = workflow.projectDir
    email_fields['summary'] = summary
    email_fields['summary']['Date Started'] = workflow.start
    email_fields['summary']['Date Completed'] = workflow.complete
    email_fields['summary']['Pipeline script file path'] = workflow.scriptFile
    email_fields['summary']['Pipeline script hash ID'] = workflow.scriptId
    if (workflow.repository) email_fields['summary']['Pipeline repository Git URL'] = workflow.repository
    if (workflow.commitId) email_fields['summary']['Pipeline repository Git Commit'] = workflow.commitId
    if (workflow.revision) email_fields['summary']['Pipeline Git branch/tag'] = workflow.revision
    if (workflow.container) email_fields['summary']['Docker image'] = workflow.container
    email_fields['summary']['Nextflow Version'] = workflow.nextflow.version
    email_fields['summary']['Nextflow Build'] = workflow.nextflow.build
    email_fields['summary']['Nextflow Compile Timestamp'] = workflow.nextflow.timestamp

    // TODO nf-core: If not using MultiQC, strip out this code (including params.maxMultiqcEmailFileSize)
    // On success try attach the multiqc report
    def mqc_report = null
    try {
        if (workflow.success) {
            mqc_report = multiqc_report.getVal()
            if (mqc_report.getClass() == ArrayList) {
                log.warn "[nf-core/hgtsimulation] Found multiple reports from process 'multiqc', will use only one"
                mqc_report = mqc_report[0]
            }
        }
    } catch (all) {
        log.warn "[nf-core/hgtsimulation] Could not attach MultiQC report to summary email"
    }

    // Check if we are only sending emails on failure
    email_address = params.email
    if (!params.email && params.email_on_fail && !workflow.success) {
        email_address = params.email_on_fail
    }

    // Render the TXT template
    def engine = new groovy.text.GStringTemplateEngine()
    def tf = new File("$baseDir/assets/email_template.txt")
    def txt_template = engine.createTemplate(tf).make(email_fields)
    def email_txt = txt_template.toString()

    // Render the HTML template
    def hf = new File("$baseDir/assets/email_template.html")
    def html_template = engine.createTemplate(hf).make(email_fields)
    def email_html = html_template.toString()

    // Render the sendmail template
    def smail_fields = [ email: email_address, subject: subject, email_txt: email_txt, email_html: email_html, baseDir: "$baseDir", mqcFile: mqc_report, mqcMaxSize: params.maxMultiqcEmailFileSize.toBytes() ]
    def sf = new File("$baseDir/assets/sendmail_template.txt")
    def sendmail_template = engine.createTemplate(sf).make(smail_fields)
    def sendmail_html = sendmail_template.toString()

    // Send the HTML e-mail
    if (email_address) {
        try {
          if ( params.plaintext_email ){ throw GroovyException('Send plaintext e-mail, not HTML') }
          // Try to send HTML e-mail using sendmail
          [ 'sendmail', '-t' ].execute() << sendmail_html
          log.info "[nf-core/hgtsimulation] Sent summary e-mail to $email_address (sendmail)"
        } catch (all) {
          // Catch failures and try with plaintext
          [ 'mail', '-s', subject, email_address ].execute() << email_txt
          log.info "[nf-core/hgtsimulation] Sent summary e-mail to $email_address (mail)"
        }
    }

    // Write summary e-mail HTML to a file
    def output_d = new File( "${params.outdir}/pipeline_info/" )
    if (!output_d.exists()) {
      output_d.mkdirs()
    }
    def output_hf = new File( output_d, "pipeline_report.html" )
    output_hf.withWriter { w -> w << email_html }
    def output_tf = new File( output_d, "pipeline_report.txt" )
    output_tf.withWriter { w -> w << email_txt }

    c_reset = params.monochrome_logs ? '' : "\033[0m";
    c_purple = params.monochrome_logs ? '' : "\033[0;35m";
    c_green = params.monochrome_logs ? '' : "\033[0;32m";
    c_red = params.monochrome_logs ? '' : "\033[0;31m";

    if (workflow.stats.ignoredCount > 0 && workflow.success) {
      log.info "${c_purple}Warning, pipeline completed, but with errored process(es) ${c_reset}"
      log.info "${c_red}Number of ignored errored process(es) : ${workflow.stats.ignoredCount} ${c_reset}"
      log.info "${c_green}Number of successfully ran process(es) : ${workflow.stats.succeedCount} ${c_reset}"
    }

    if (workflow.success) {
        log.info "${c_purple}[nf-core/hgtsimulation]${c_green} Pipeline completed successfully${c_reset}"
    } else {
        checkHostname()
        log.info "${c_purple}[nf-core/hgtsimulation]${c_red} Pipeline completed with errors${c_reset}"
    }

}


def nfcoreHeader(){
    // Log colors ANSI codes
    c_reset = params.monochrome_logs ? '' : "\033[0m";
    c_dim = params.monochrome_logs ? '' : "\033[2m";
    c_black = params.monochrome_logs ? '' : "\033[0;30m";
    c_green = params.monochrome_logs ? '' : "\033[0;32m";
    c_yellow = params.monochrome_logs ? '' : "\033[0;33m";
    c_blue = params.monochrome_logs ? '' : "\033[0;34m";
    c_purple = params.monochrome_logs ? '' : "\033[0;35m";
    c_cyan = params.monochrome_logs ? '' : "\033[0;36m";
    c_white = params.monochrome_logs ? '' : "\033[0;37m";

    return """    -${c_dim}--------------------------------------------------${c_reset}-
                                            ${c_green},--.${c_black}/${c_green},-.${c_reset}
    ${c_blue}        ___     __   __   __   ___     ${c_green}/,-._.--~\'${c_reset}
    ${c_blue}  |\\ | |__  __ /  ` /  \\ |__) |__         ${c_yellow}}  {${c_reset}
    ${c_blue}  | \\| |       \\__, \\__/ |  \\ |___     ${c_green}\\`-._,-`-,${c_reset}
                                            ${c_green}`._,._,\'${c_reset}
    ${c_purple}  nf-core/hgtsimulation v${workflow.manifest.version}${c_reset}
    -${c_dim}--------------------------------------------------${c_reset}-
    """.stripIndent()
}

def checkHostname(){
    def c_reset = params.monochrome_logs ? '' : "\033[0m"
    def c_white = params.monochrome_logs ? '' : "\033[0;37m"
    def c_red = params.monochrome_logs ? '' : "\033[1;91m"
    def c_yellow_bold = params.monochrome_logs ? '' : "\033[1;93m"
    if (params.hostnames) {
        def hostname = "hostname".execute().text.trim()
        params.hostnames.each { prof, hnames ->
            hnames.each { hname ->
                if (hostname.contains(hname) && !workflow.profile.contains(prof)) {
                    log.error "====================================================\n" +
                            "  ${c_red}WARNING!${c_reset} You are running with `-profile $workflow.profile`\n" +
                            "  but your machine hostname is ${c_white}'$hostname'${c_reset}\n" +
                            "  ${c_yellow_bold}It's highly recommended that you use `-profile $prof${c_reset}`\n" +
                            "============================================================"
                }
            }
        }
    }
}
